bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AbstractRetryingOperation.java

```diff
@@ -336,7 +337,12 @@ protected void resetStatusBasedBackoff() {
    * @return a {@link ExponentialRetryAlgorithm} object.
    */
   private ExponentialRetryAlgorithm createRetryAlgorithm(ApiClock clock) {
-    int timeoutMs = retryOptions.getMaxElapsedBackoffMillis();
+    long timeoutMs = retryOptions.getMaxElapsedBackoffMillis();
+
+    Deadline deadline = getOperationCallOptions().getDeadline();
```

Comment:
This is a bit off...deadline should be overall timeout, rather then just the retry amount. Please add a TODO to fix it

Reply:
how does this interact with grpc deadline? I'm concerned about this following case:
- deadline is set to 10s
- first rpc returns in 9s with a retryable error
- the retry mechanism schedules a retry in 2s
- 2nd rpc returns a local deadline exceeded
- the retry mechanism reschedules
- 3rd rpc returns a local deadline exceeded
...
until 10s pass after the first rpc failed

Reply:
The `CallOptions` are sent for each RPC.  If a deadline is set, the retry mechanism will only go as long as the Deadline in the `CallOptions`.

Something similar to your case:

- deadline is set to 10s
- first rpc returns in 9s with a retryable error
- The retry schedules a timeout in 10s - 9s = 1s, with an initial retry of a few ms.
- The second RPC still has roughly 1s to complete.
- 1s goes by, DEADLINE_EXCEED happens due to the CallOptions
- Retry logic sees that Timeout occurred, and stops retries.